### PR TITLE
mgr/dashboard: Select any object gateway on local cluster

### PR DESCRIPTION
**Pull request by alfonsomthd**
WIP

Fixes: https://tracker.ceph.com/issues/47375

* Settings:
  - Refactoring: now accepting more than 1 type of value.
  - RGW_API_ACCESS_KEY & RGW_API_SECRET_KEY accept string (backward compatibility: legacy behavior) as well as dictionary of strings for connecting multiple daemons.
Screenshot from 2020-11-18 17-14-48
  - Ease of use: deprecated: mgr/dashboard/RGW_API_USER_ID: not useful anymore (kept for backward compatibility).

* UI/UX:
  - Created context component (to be shown only on rgw-related routes) for selecting operating daemon.
  - Daemon selector only shown if there is more than 1 daemon running on local cluster (to reduce cognitive load).

Screenshot from 2021-02-19 16-02-47

HOW TO TEST (to be improved):
1) Configure RGW multisite on your local cluster and start at least 2 daemon running on different realms.
2) Set access/secret keys from system rgw users in Dashboard settings.

## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

**Review by epuertat — COMMENTED** [review]
Awesome job, @alfonsomthd! We finally have support for multiple RGW daemons!

I have code-reviewed and done some manual testing with ceph-dev and a multi-cluster setup (x2 RGW).

Some comments on the UI aspects:
* "Current Obj. Gateway": I'd either go for the complete version ("Selected Object Gateway"), or the fully abbreviated ("Selected RGW Daemon"). I'd put " "selected" vs. "current" since it emphasizes the fact that it's user-driven.
* On the layout, I understand that the breadcrumb service makes it not simply to do that, but there it goes my proposal:
  * Current: 
  image
  * Proposed:
  image
* I miss the context selector not being more eye-catching (surrounding it with a dark frame to make it more evident its selectable nature) and improve the visual interaction feedback (using the clickable mousepointer on mouse over rather that the text-insert one), like the [Patternfly one](https://www.patternfly.org/v4/components/context-selector):

dropdown

dropdown2

What do you think?

File: src/pybind/mgr/dashboard/settings.py [hunk 1]
```diff
@@ -3,12 +3,47 @@
...
+                if setting_type.__name__ == 'bool' and str(value).lower() == 'false':
+                    return False
+                elif setting_type.__name__ == 'dict':
+                    return literal_eval(value)
+                return setting_type(value)
+            except (SyntaxError, TypeError, ValueError) as error:
+                if type_index == len(self.types) - 1:
+                    raise error
+
+
```
[the remark this answers is not in the document]
Comment (alfonsomthd) [reply src/pybind/mgr/dashboard/settings.py:46]: Yeah, better to do that cleanup in its own PR.

**Review by epuertat — APPROVED** [review]

**Comment by alfonsomthd** [comment]
jenkins test docs

File: src/pybind/mgr/dashboard/frontend/src/app/core/context/context.component.html [hunk 2]
```diff
@@ -0,0 +1,26 @@
+<ng-container *ngIf="{ ftMap: featureToggleMap$ | async, daemons: rgwDaemonService.daemons$ | async, selectedDaemon: rgwDaemonService.selectedDaemon$ | async } as data">
```
[the remark this answers is not in the document]
Comment (alfonsomthd) [reply src/pybind/mgr/dashboard/frontend/src/app/core/context/context.component.html:1]: After conversation in daily standup, we agree that these kind of improvements can be done in a follow-up PR.

**Comment by alfonsomthd** [comment]
> What do you think?

@epuertat I've updated screenshot in PR description to reflect the selector Look'n'Feel agreement achieved during daily standup.

**Review by nizamial09 — APPROVED** [review]
Awesome!

File: src/pybind/mgr/dashboard/frontend/src/app/shared/services/timer.service.ts [hunk 3]
```diff
@@ -21,6 +21,7 @@ export class TimerService {
     return timer(dueTime, refreshInterval, this.ngZone.leave).pipe(
       observeOn(this.ngZone.enter),
       switchMap(next),
+      retry(),
```
Comment (avanthakkar) [inline src/pybind/mgr/dashboard/frontend/src/app/shared/services/timer.service.ts:24]: Not sure if we really need this @alfonsomthd ? Everything else LGTM!

**Review by avanthakkar — APPROVED** [review]

File: src/pybind/mgr/dashboard/frontend/src/app/shared/services/timer.service.ts (hunk 3 above)
Comment (alfonsomthd) [reply src/pybind/mgr/dashboard/frontend/src/app/shared/services/timer.service.ts:24]: I agree that resubscription behaviour does not apply here, so I removed it. If needed, it should be implemented in wrapped function (```next```).

**Review by jmolmo — APPROVED** [review]

**Pull request merged**